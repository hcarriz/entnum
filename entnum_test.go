package entnum_test

import (
	"io/fs"
	"os"
	"path/filepath"
	"testing"

	"entgo.io/ent/entc"
	"entgo.io/ent/entc/gen"
	"github.com/gkampitakis/go-snaps/snaps"
	"github.com/hcarriz/entnum"
)

func TestEntc(t *testing.T) {

	type test struct {
		name    string
		args    []entnum.Option
		wantErr bool
	}

	tests := []test{
		{
			name:    "generate with entc",
			args:    []entnum.Option{},
			wantErr: false,
		},
		{
			name: "custom verb",
			args: []entnum.Option{
				entnum.Name("return"),
			},
			wantErr: false,
		},
		{
			name: "missing verb",
			args: []entnum.Option{
				entnum.Name(""),
			},
			wantErr: true,
		},
		{
			name: "verb that starts with number",
			args: []entnum.Option{
				entnum.Name("4you"),
			},
			wantErr: true,
		},
		{
			name: "verb that starts with punctuation",
			args: []entnum.Option{
				entnum.Name("!you"),
			},
			wantErr: true,
		},
	}

	for _, tt := range tests {

		t.Run(tt.name, func(t *testing.T) {

			dir := t.TempDir()

			pkg := "ent"
			sch := "schema"
			result := "entnum.go"
			schmaPath := "./test/schema"

			config := &gen.Config{
				Target:  filepath.Join(dir, pkg),
				Package: pkg,
				Schema:  sch,
			}

			got, err := entnum.New(tt.args...)
			if (err != nil) != tt.wantErr {
				t.Errorf("New() error = %v, wantErr %v", err, tt.wantErr)
				return
			}
			if err != nil && tt.wantErr {
				return
			}

			opts := []entc.Option{
				entc.Extensions(got),
			}

			if err := entc.Generate(schmaPath, config, opts...); err != nil {
				t.Fatalf("running ent codegen: %v", err)
				return
			}

			f := os.DirFS(dir)

			data, err := fs.ReadFile(f, filepath.Join(pkg, result))
			if err != nil {
				t.Errorf("unable to open directory: %v", err)
				return
			}

			snaps.MatchSnapshot(t, string(data))

		})
	}

}

// func TestNew(t *testing.T) {
// 	type args struct {
// 		opts []Option
// 	}
// 	tests := []struct {
// 		name    string
// 		args    args
// 		want    *Extension
// 		wantErr bool
// 	}{
// 		// TODO: Add test cases.
// 	}
// 	for _, tt := range tests {
// 		t.Run(tt.name, func(t *testing.T) {
// 			got, err := New(tt.args.opts...)
// 			if (err != nil) != tt.wantErr {
// 				t.Errorf("New() error = %v, wantErr %v", err, tt.wantErr)
// 				return
// 			}
// 			if !reflect.DeepEqual(got, tt.want) {
// 				t.Errorf("New() = %v, want %v", got, tt.want)
// 			}
// 		})
// 	}
// }
