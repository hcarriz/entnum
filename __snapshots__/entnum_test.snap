
[TestEntc/generate_with_entc - 1]
// Code generated by ent, DO NOT EDIT.

package ent

import (
    "ent/todo"
)

// AllTodoStatus will return all of todo's status values.
func AllTodoStatus() []todo.Status {
    return []todo.Status{
        todo.StatusOngoing,
        todo.StatusCompleted,
        todo.StatusLater,
    }
}

---

[TestEntc/custom_verb - 1]
// Code generated by ent, DO NOT EDIT.

package ent

import (
    "ent/todo"
)

// ReturnTodoStatus will return all of todo's status values.
func ReturnTodoStatus() []todo.Status {
    return []todo.Status{
        todo.StatusOngoing,
        todo.StatusCompleted,
        todo.StatusLater,
    }
}

---
